bin/respirate

```diff
@@ -5,6 +5,11 @@ require_relative "../loader"
 
 d = Scheduling::Dispatcher.new
 
+if Config.heartbeat_url && Strand[prog: "Heartbeat"].nil?
+  puts "Starting heartbeat prog"
+  Strand.create_with_id(schedule: Time.now, prog: "Heartbeat", label: "wait")
```

Comment:
Not sure if this is good place to create a strand. Maybe we should create the Strand manually.

Reply:
When we deploy to a new location, I aim to reduce the need for manual operations. It can be easily forgotten. I don't feel very strong about it, but I dislike manually creating strands. Given that the VmHost model has some operational helper functions, we could add an ﻿`assemble` function to `﻿Prog::Heartbeat`.

Reply:
I agree with enes on the matter of automation here, but I'm a bit confused how this prog is supposed to better inform us of issues than hooking up the health status to assessing `pg_stat_activity` table or, more expensively, inspecting the Strand schedules. Alternatively, we can use logging to identify a lot of seriously behind schedule strands as they execute. This would allow us to identify brownouts better, and it may even come out with less code. I'll experiment if I have time after reviewing a couple more DNS implementations.

Reply:
It might not be better, and here are my reasons.
We are already using StatusCake for monitoring web server health. We can use the same third-party service to monitor respirate's health, thus avoiding the introduction of a new component.

We need to run "pg_stat_activity" on another third-party service, which requires us to ensure the availability of that service as well.

Additionally, we could implement both monitoring systems to increase certainty.

Reply:
I believe we need to incorporate metrics into our service. Metrics provide a valuable resource for generating pages.

Reply:
Right, but statuscake monitors by calling a GET request, whereas this heartbeat makes a GET: that's pretty different, isn't it? I'm suggesting a `/health` endpoint that does stuff that's cheap enough to be unauthenticated (since I doubt we can use firewalls), or alternatively, `/obscure-random-string` to do the same, mitigating DOS issues.

I think a static /health is a good idea for development and small installations though, just makes things easier.

Reply:
This health check is for respirate. It's the reverse of typical monitoring. It's a "Push Monitor" at StatusCake. We call  StatusCake, instead of StatusCake call us. Would you recommend adding a web router to the respirate dyno and serving the `﻿/health` endpoint?

Reply:
No, I'd not suggest that (and Heroku doesn't allow it, right?) but I think it's quite alright: by reading `pg_stat_activity.application_name` and/or `now() - strand.schedule order by limit $small` from a web process, you could do everything that needs to be done for quite some time: checking database connectivity, and checking timely checking progression of work. 

Even if the web dyno is inferring the `respirate` situation by doing so. The latter type of query also can signal brownouts, where, say, a bunch of `respirate` activity is backing up, for any reason: a sudden spike in use of expensive labels, problems with some `respirate` nodes, etc.

I kinda feel like the dial-out is an interesting but finicky feature, not in as common circulation in industry. I suppose we could always invert the logic of a public endpoint and not send the heartbeat when we don't meet our criteria about `strand.schedule` distribution or `pg_stat_activity.application_name` counts.

Reply:
Ya know, maybe let's use the heartbeat URL, but let's only heartbeat in a more stringent condition. I'll submit a modification.

Reply:
> Right, but statuscake monitors by making a GET request, as-is it goes to one of our regular unauthenticated HTTP web pages

Not related with the discussion, but for the record; StatusCake authenticates to our API. I created a user for StatusCake for this purpose.